# Estruturas de dados

## Lista Sequencial

In [6]:
import numpy as np
import numpy.typing as npt
from collections.abc import MutableSequence

class ListaSequencial:
    def __init__(self, capacidade: int):
        self.capacidade: int = capacidade
        self.valores: npt.NDArray[np.generic[int]] = np.empty(self.capacidade, dtype=int)
        self.ultima_posicao: int = -1

    def insere(self, valor: int):
        if self.ultima_posicao == self.capacidade - 1:
            print('Capacidade máxima atingida')
        else:
            self.ultima_posicao += 1
            self.valores[self.ultima_posicao] = valor

    def pesquisar(self, valor: int):
        for i in range(0, self.ultima_posicao + 1):
            if self.valores[i] == valor:
                return i
        return -1

    def excluir(self, valor: int):
        posicao = self.pesquisar(valor)
        if posicao == -1:
            return -1
        else:
            for i in range(posicao, self.ultima_posicao):
                self.valores[i] = self.valores[i + 1]
            self.ultima_posicao -= 1

    def imprime(self):
        if self.ultima_posicao == -1:
            print("O vetor está vazio")
        else:
            for i in range(self.ultima_posicao + 1):
                print(f"{i} - {self.valores[i]}")

## Ordenação

In [7]:
def bubble_sort(lista: MutableSequence[int]) -> MutableSequence[int]:
    n = len(lista)

    for i in range(n - 1):
        for j in range(n - i - 1):
            if lista[j] > lista[j + 1]:
                lista[j], lista[j + 1] = lista[j + 1], lista[j]
    return lista

def selection_sort(lista: MutableSequence[int]) -> MutableSequence[int]:
    n = len(lista)

    for i in range(n):
        smallest_idx = i
        for j in range(i, n):
            if lista[j] < lista[smallest_idx]:
                smallest_idx = j
        if smallest_idx != i:
            lista[i], lista[smallest_idx] = lista[smallest_idx], lista[i]
    return lista


def choose_pivot(arr: MutableSequence[int], start: int, end: int) -> tuple[int, int]:
    idx = (end + start) // 2

    return arr[idx], idx

def swap(arr: MutableSequence[int], i: int, j: int):
    arr[i], arr[j] = arr[j], arr[i]

def partition(arr: MutableSequence[int], start: int, end: int) -> int:
    i = start - 1
    pivot, pivot_idx = choose_pivot(arr, start, end)

    swap(arr, pivot_idx, end)
    for j in range(start, end):
        if arr[j] < pivot:
            i += 1
            swap(arr, i, j)
    i += 1
    swap(arr, i, end)
    return i

def quick_sort(arr: MutableSequence[int], start: int | None = None, end: int | None = None):
    if start is None:
        start = 0
    if end is None:
        end = len(arr) - 1

    if start >= end:
        return
    pi = partition(arr, start, end)
    quick_sort(arr, start, pi - 1)
    quick_sort(arr, pi + 1, end)


## Fila

In [8]:
class Fila:
    def __init__(self, capacidade: int):
        self.capacidade: int = capacidade
        self.inicio: int = 0
        self.final: int = -1
        self.numero_elementos: int = 0
        self.valores: npt.NDArray[np.int_]  = np.empty(self.capacidade, dtype=np.int_)

    def fila_vazia(self):
        return self.numero_elementos == 0

    def fila_cheia(self):
        return self.numero_elementos == self.capacidade

    def enfileirar(self, valor: int):
        if self.fila_cheia():
            print('A fila está cheia')
            return
        if self.final == self.capacidade - 1:
            self.final = -1
        self.final += 1
        self.valores[self.final] = valor
        self.numero_elementos += 1

    def desenfileirar(self):
        if self.fila_vazia():
            print('A fila já está vazia')
            return
        temp = self.valores.item(self.inicio)
        self.inicio += 1
        if self.inicio == self.capacidade - 1:
            self.inicio = 0
        else:
            self.numero_elementos -= 1
            return temp

    def primeiro(self):
        if self.fila_vazia():
            return -1
        else:
            return self.valores.item(self.inicio)

    def visualizar(self):
      print('Tamanho da Fila..: ', self.capacidade)
      print('Início da fila...: ', self.inicio)
      print('Final da fila....: ', self.final)
      print('Variável Nu. El..:', self.numero_elementos)

    def listageral(self):
      for i in range(self.capacidade):
        print(self.valores.item(i), end=' ')

## Pilha

In [9]:
class Pilha:

    def __init__(self):
        self.items: list[int] = []

    def empilhar(self, item: int):
        self.items = self.items + [item]

    def desempilhar(self):
        if not self.is_vazia():
            item_removido = self.items[-1]
            self.items = self.items[:-1]
            return item_removido
        else:
            print("A pilha está vazia. Não é possível desempilhar.")

    def topo(self):
        if not self.is_vazia():
            return self.items[-1]
        else:
            print("A pilha está vazia. Não há topo para visualizar.")

    def is_vazia(self):
        return len(self.items) == 0

    def tamanho(self):
        return len(self.items)

    def imprimir(self):
        if not self.is_vazia():
            print("Itens da pilha:")
            for item in self.items:
                print(item)
        else:
            print("A pilha está vazia. Não há itens para imprimir.")

## Exemplo para calcular o tempo de execução

In [10]:
import time
import random

vetor = [random.randint(1, 100) for _ in range(100)]

# Tempo de Execução Bubble Sort
inicio = time.time()
resultado = bubble_sort(vetor)
fim = time.time()
tempo_execucao = fim - inicio
print(tempo_execucao)

0.000217437744140625


# Importando dados do GitHub

In [11]:
import pandas as pd

# URL do arquivo csv no GitHub
url = 'https://raw.githubusercontent.com/nisston/disciplinaestruturadedados/main/data_temp.csv'

# Ler o arquivo CSV usando pandas
df1 = pd.read_csv(url)

In [12]:
# Exibi o tamanho do dataframe
df1.shape

(15907, 3)

In [13]:
# Exibir a estrutura do DataFrame
df1.info()

<class 'pandas.DataFrame'>
RangeIndex: 15907 entries, 0 to 15906
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   key          15907 non-null  int64  
 1   time         15907 non-null  str    
 2   Temperature  13006 non-null  float64
dtypes: float64(1), int64(1), str(1)
memory usage: 372.9 KB


In [14]:
# Exibi todos os registro com campos em branco
df1.isnull().sum()

key               0
time              0
Temperature    2901
dtype: int64

In [15]:
# Eliminando registros em branco
df1.dropna(inplace=True, axis=0)

In [16]:
df1.columns

Index(['key', 'time', 'Temperature'], dtype='str')

In [17]:
# Temperatura em Kelvin (Celsius e Fahrenheit)
df1['Temperature'][:6]

0    1571.0
1    1604.0
2    1618.0
3    1601.0
4    1613.0
5    1581.0
Name: Temperature, dtype: float64

# Trabalhando com intervalos de dados

In [18]:
# Pegando os 5 primeiros registros
df1[0:5]

,key,time,Temperature
0,1,2019-05-03 11:16:18,1571.0
1,1,2019-05-03 11:25:53,1604.0
2,1,2019-05-03 11:29:11,1618.0
3,1,2019-05-03 11:30:01,1601.0
4,1,2019-05-03 11:30:39,1613.0


In [19]:
# Pegando os 5 primeiros registros
df1[:5]

,key,time,Temperature
0,1,2019-05-03 11:16:18,1571.0
1,1,2019-05-03 11:25:53,1604.0
2,1,2019-05-03 11:29:11,1618.0
3,1,2019-05-03 11:30:01,1601.0
4,1,2019-05-03 11:30:39,1613.0


In [20]:
# Pegando os registros de 5 até 14
df1[5:15]

,key,time,Temperature
5,2,2019-05-03 11:37:27,1581.0
6,2,2019-05-03 11:38:00,1577.0
7,2,2019-05-03 11:49:38,1589.0
8,2,2019-05-03 11:55:50,1604.0
9,2,2019-05-03 11:58:24,1608.0
10,2,2019-05-03 11:59:12,1602.0
11,3,2019-05-03 12:13:17,1596.0
12,3,2019-05-03 12:18:36,1597.0
13,3,2019-05-03 12:25:06,1598.0
14,3,2019-05-03 12:30:48,1599.0


In [21]:
# Pegando os 5 últimos registros
df1[-5:]

,key,time,Temperature
15878,3237,2019-08-31 22:44:04,1569.0
15883,3238,2019-08-31 23:30:31,1584.0
15886,3239,2019-09-01 01:31:47,1598.0
15894,3240,2019-09-01 02:39:01,1617.0
15902,3241,2019-09-01 04:03:30,1586.0


In [22]:
# Pegando um intervalo de registro com colunas específicas
df1[:4][['key','Temperature']]

,key,Temperature
0,1,1571.0
1,1,1604.0
2,1,1618.0
3,1,1601.0


In [23]:
# Como converter um intervalo de registro de uma coluna para uma lista
dfprovaa = df1[:1000]['Temperature'].tolist()
dfprovaa

[1571.0,
 1604.0,
 1618.0,
 1601.0,
 1613.0,
 1581.0,
 1577.0,
 1589.0,
 1604.0,
 1608.0,
 1602.0,
 1596.0,
 1597.0,
 1598.0,
 1599.0,
 1599.0,
 1601.0,
 1623.0,
 1625.0,
 1576.0,
 1602.0,
 1543.0,
 1581.0,
 1600.0,
 1600.0,
 1596.0,
 1586.0,
 1582.0,
 1599.0,
 1595.0,
 1599.0,
 1577.0,
 1604.0,
 1600.0,
 1598.0,
 1587.0,
 1589.0,
 1597.0,
 1589.0,
 1592.0,
 1574.0,
 1580.0,
 1585.0,
 1586.0,
 1586.0,
 1586.0,
 1588.0,
 1593.0,
 1616.0,
 1608.0,
 1583.0,
 1597.0,
 1606.0,
 1606.0,
 1585.0,
 1588.0,
 1584.0,
 1595.0,
 1591.0,
 1596.0,
 1602.0,
 1610.0,
 1613.0,
 1621.0,
 1615.0,
 1619.0,
 1583.0,
 1581.0,
 1602.0,
 1596.0,
 1600.0,
 1606.0,
 1605.0,
 1607.0,
 1611.0,
 1613.0,
 1599.0,
 1598.0,
 1581.0,
 1590.0,
 1609.0,
 1607.0,
 1611.0,
 1611.0,
 1596.0,
 1591.0,
 1599.0,
 1599.0,
 1588.0,
 1600.0,
 1593.0,
 1592.0,
 1604.0,
 1598.0,
 1595.0,
 1625.0,
 1616.0,
 1610.0,
 1605.0,
 1603.0,
 1598.0,
 1604.0,
 1597.0,
 1582.0,
 1584.0,
 1608.0,
 1597.0,
 1603.0,
 1603.0,
 1579.0,
 1581.0,
 

In [24]:
type(dfprovaa)

list

In [25]:
# Tamanho da lista
len(dfprovaa)

1000

# Questões

01 - Após carregar a base de dados, classifique a temperatura utilizando o método bubble_sort e depois apresente as 15 primeiras temperaturas mais altas.


In [31]:
temperaturas = df1['Temperature'].tolist()
temperaturas = bubble_sort(temperaturas)
minimas = temperaturas[-15:]
print(f"{minimas=}\n{len(minimas)=}")

minimas=[1684.0, 1684.0, 1684.0, 1684.0, 1684.0, 1684.0, 1690.0, 1690.0, 1691.0, 1694.0, 1696.0, 1700.0, 1702.0, 1704.0, 1705.0]
len(minimas)=15


02 - Após carregar a base de dados, classifique a temperatura utilizando o método quick sort e depois apresente as 20 primeiras temperaturas mais baixa.


In [30]:
temperaturas = df1['Temperature'].tolist()
quick_sort(temperaturas)
minimas = temperaturas[:20]
print(f"{minimas=}\n{len(minimas)=}")

minimas=[1191.0, 1204.0, 1208.0, 1218.0, 1227.0, 1383.0, 1383.0, 1519.0, 1520.0, 1520.0, 1525.0, 1525.0, 1526.0, 1526.0, 1526.0, 1526.0, 1527.0, 1527.0, 1528.0, 1529.0]
len(minimas)=20


03 - Após carregar a base de dados, classifique a temperatura utilizando o método select sort e depois apresente a soma das 10 primeiras temperaturas mais altas.


In [32]:
temperaturas = df1['Temperature'].tolist()
temperaturas = selection_sort(temperaturas)
maximas = temperaturas[:10]
print(f"{maximas=}\n{sum(maximas)=}")

maximas=[1191.0, 1204.0, 1208.0, 1218.0, 1227.0, 1383.0, 1383.0, 1519.0, 1520.0, 1520.0]
sum(maximas)=13373.0


04 - Após carregar a base de dados, qual dos métodos de classifição (bubble_sort, select_sort e quick_sort) é o mais eficiente?


In [38]:
from typing import Callable
from functools import wraps

def metrificar(fn: Callable[[MutableSequence[int]], MutableSequence[int] | None]):
    @wraps(fn)
    def wrap(lista: MutableSequence[int]):
        inicio = time.time()
        _ = fn(lista) # Não interessa agora
        fim = time.time()
        return fim - inicio
    return wrap

bsort = metrificar(bubble_sort)
ssort = metrificar(selection_sort)
qsort = metrificar(quick_sort)
populacao_testes = list(range(200))
testes = [random.sample(populacao_testes, i) for i in (1, 5, 10, 100, 200)]

for teste in testes:
    winner = "Ninguém"
    maior_velocidade = float('inf')
    bsorted, ssorted, qsorted = bsort(teste), ssort(teste), qsort(teste)

    if bsorted < maior_velocidade:
        winner = 'Bubble Sort'
        maior_velocidade = bsorted
    if ssorted < maior_velocidade:
        winner = 'Selection Sort'
        maior_velocidade = ssorted
    if qsorted < maior_velocidade:
        winner = 'Quick Sort'
    print(f"{teste=}\n{winner=} tempos=bsort={bsorted:.6f}, ssort={ssorted:.6f}, qsort={qsorted:.6f}\n")

teste=[111]
winner='Quick Sort' tempos=bsort=0.000001, ssort=0.000001, qsort=0.000001

teste=[65, 69, 82, 88, 113]
winner='Selection Sort' tempos=bsort=0.000002, ssort=0.000001, qsort=0.000004

teste=[7, 47, 51, 58, 63, 87, 92, 95, 126, 193]
winner='Selection Sort' tempos=bsort=0.000003, ssort=0.000002, qsort=0.000003

teste=[1, 3, 4, 7, 9, 12, 13, 14, 19, 20, 21, 23, 25, 28, 30, 32, 34, 37, 38, 39, 43, 48, 50, 55, 56, 61, 62, 63, 64, 65, 66, 67, 71, 74, 76, 78, 81, 83, 85, 86, 87, 88, 89, 90, 91, 92, 93, 96, 98, 99, 100, 101, 102, 104, 106, 107, 109, 110, 111, 112, 113, 115, 116, 117, 118, 124, 125, 130, 134, 135, 136, 137, 138, 139, 144, 149, 151, 155, 158, 159, 162, 166, 168, 169, 170, 171, 172, 174, 176, 180, 181, 182, 183, 184, 185, 186, 193, 194, 196, 199]
winner='Quick Sort' tempos=bsort=0.000164, ssort=0.000076, qsort=0.000034

teste=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,

05 - Após carregar a base de dados, classifique os dados com o método  quick_sort, em seguida construa uma estrutura de pilha e adicione as 500 temperaturas mais altas, saltando de 2 em 2, após isso, desempilhe 50 e responda qual é o topo da pilha?


In [39]:
temperaturas: list[int] = df1['Temperature'].tolist()
quick_sort(temperaturas)
temperaturas = temperaturas[-500::2]
pilha = Pilha()
for temperatura in temperaturas:
    pilha.empilhar(temperatura)
for _ in range(50):
    _ = pilha.desempilhar()
print(f"Topo da pilha: {pilha.topo()}")

Topo da pilha: 1659.0


06 - Após carregar a base de dados, classifique os dados com o método select_sort, em seguida construa uma estrutura de pilha e adicione (empilhar) as 400 temperaturas mais altas, desempilhe 100 e responda que esta no topo da pilha?

In [40]:
temperaturas: MutableSequence[int] = df1['Temperature'].tolist()
temperaturas = selection_sort(temperaturas)
temperaturas = temperaturas[-400:]
pilha = Pilha()
for temperatura in temperaturas:
    pilha.empilhar(temperatura)
for _ in range(100):
    _ = pilha.desempilhar()
print(f"Topo da pilha: {pilha.topo()}")

Topo da pilha: 1659.0


07 - Após carregar a base de dados, classifique os dados com o método  quick_sort, em seguida construa uma estrutura de pilha e adicione as 100 temperaturas mais altas (empilhar), após isso, desempilhe 30 e responda ao que se pede:

1.   Quem tá no topo da pilha?
2.   Qual é a soma das 10 temperaturas desempilhadas?
3.   Qual a média das temperaturas desempilhadas?
4.   Qual o maior valor das temperaturas desempilhadas?



In [41]:
temperaturas = df1['Temperature'].tolist()
quick_sort(temperaturas)
temperaturas: list[int] = temperaturas[-100:]
pilha = Pilha()
for temperatura in temperaturas:
    pilha.empilhar(temperatura)
desempilhadas: list[int] = []
for _ in range(30):
    desempilhada = pilha.desempilhar()

    if desempilhada is not None:
        desempilhadas.append(desempilhada)
print(f"Topo da pilha: {pilha.topo()}")
print(f"Soma das temperaturas desempilhadas: {sum(desempilhadas)}")
print(f"Média das temperaturas desempilhadas: {sum(desempilhadas) / len(desempilhadas)}")
print(f"Maior valor das temperaturas desempilhadas: {max(desempilhadas)}")

Topo da pilha: 1676.0
Soma das temperaturas desempilhadas: 50568.0
Média das temperaturas desempilhadas: 1685.6
Maior valor das temperaturas desempilhadas: 1705.0


08 - Após carregar a base de dados, classifique os dados com o método  select_sort, em seguida construa uma estrutura de fila e adicione as 100 temperaturas mais altas (enfileirar), após isso, desemfileire 30 e responda ao que se pede:

1.   Quem tá no início da fila?
2.   Qual é a soma das 30 temperaturas desenfileiradas?
3.   Qual a média das temperaturas desenfileiradas?
4.   Qual o maior valor das temperaturas desenfileiradas?

In [ ]:
# Seu código aqui

09 - Após carregar a base de dados, construa uma estrutura de pilha e adicione as 100 primeiras temperaturas (0-99), após isso, crie uma estrutura de fila e enfileire as 30 temperaturas da estrutura de pilha, ao final desenfileirar 15 e apresentar quem é a primeira temperatura da fila.

In [ ]:
# Seu código aqui

10 - Após carregar a base de dados, construa uma estrutura de pilha e adicione as 200 últimas temperaturas, em seguida desempilhe 50 e apresente ao final a temperatura que ficou no topo da pilha além da soma e média das temperaturas que foram desempilhadas.

In [ ]:
# Seu código aqui

11 - Após carregar a base de dados, crie duas filas (A e B), a primeira para enfileirar os 200 primeiros registros da base (0:199), e na segunda, você defe enfileirar os 200 últimos registro da base (-200:).

In [ ]:
# Seu código aqui

12 - Após carregar a base de dados, crie uma fila, depois enfileirar os 200 primeiros registros da base (0:199), e em seguida os 200 últimos registro da base (-200:). Desenfileirar os 250 e apresentar o primeiro da fila.

In [ ]:
# Seu código aqui

13 - Após carregar a base de dados, crie uma pilha, depois empilhar os 100 primeiros registros da base (0:99), em seguida os 200 últimos registro da base (-200:) e depois desempilhar os 135 primeiros e apresentar o elemento que ficou no topo da pilha.

In [ ]:
# Seu código aqui

14 - Após carregar a base de dados, crie uma fila e uma pilha, enfileirar os 150 primeiros registros da base (0:149), em seguida empilhar os 150 registros seguintes (150:299), em seguinda desenfileirar e desempinhar os 50 elementos de cada estrutura, ao final apresente a soma deles, quando desenfileira pegar o valor e somar com o elemento será desempilhado.

In [ ]:
# Seu código aqui